src/GovUk.Education.ExploreEducationStatistics.Data.Services/SubjectMetaService.cs

```diff
@@ -32,91 +36,168 @@ private enum SubjectMetaQueryStep
+        }
+
+        [BlobCache(typeof(SubjectMetaCacheKey))]
+        private Task<SubjectMetaViewModel> GetSubjectMeta(ReleaseSubject releaseSubject, SubjectMetaCacheKey cacheKey)
+        {
+            return GetSubjectMetaViewModel(releaseSubject);
         }
 
-        public async Task<Either<ActionResult, SubjectMetaViewModel>> GetSubjectMetaRestricted(Guid subjectId)
+        public async Task<Either<ActionResult, SubjectMetaViewModel>> GetSubjectMetaRestricted(
+            Guid releaseId,
+            Guid subjectId)
         {
-            return await _persistenceHelper.CheckEntityExists<Subject>(subjectId)
+            return await CheckReleaseSubjectExists(releaseId, subjectId)
                 .OnSuccess(CheckCanViewSubjectData)
                 .OnSuccess(GetSubjectMetaViewModel);
         }
 
         public Task<Either<ActionResult, SubjectMetaViewModel>> GetSubjectMeta(
-            ObservationQueryContext query, CancellationToken cancellationToken)
+            ObservationQueryContext query,
+            CancellationToken cancellationToken)
         {
-            return _persistenceHelper.CheckEntityExists<Subject>(query.SubjectId)
-                .OnSuccess(_ => GetSubjectMetaViewModelFromQuery(query, cancellationToken));
+            return CheckSubjectExistsOnLatestPublishedVersion(query.SubjectId)
+                .OnSuccess(releaseSubject =>
+                    GetSubjectMetaViewModelFromQuery(query, releaseSubject, cancellationToken));
         }
 
         public Task<Either<ActionResult, SubjectMetaViewModel>> GetSubjectMetaRestricted(
-            ObservationQueryContext query, CancellationToken cancellationToken)
+            Guid releaseId,
+            ObservationQueryContext query,
+            CancellationToken cancellationToken)
         {
-            return _persistenceHelper.CheckEntityExists<Subject>(query.SubjectId)
+            return CheckReleaseSubjectExists(releaseId, query.SubjectId)
                 .OnSuccess(CheckCanViewSubjectData)
-                .OnSuccess(_ => GetSubjectMetaViewModelFromQuery(query, cancellationToken));
+                .OnSuccess(releaseSubject =>
+                    GetSubjectMetaViewModelFromQuery(query, releaseSubject, cancellationToken));
+        }
+
+        public async Task<Either<ActionResult, Unit>> UpdateSubjectFilters(
+            Guid releaseId,
+            Guid subjectId,
+            List<FilterUpdateViewModel> request)
+        {
+            return await CheckReleaseSubjectExists(releaseId, subjectId)
+                .OnSuccessVoid(async rs =>
+                {
+                    // TODO EES-3345 Validate all entries are present and only entries that relate to the subject
+
+                    // Set the sequence based on the order of filters, filter groups and indicators observed
+                    // in the request
+                    rs.FilterSequence = request.Select(filter =>
+                            new FilterSequenceEntry
```

Comment:
Yes, since I'm only storing the Id's I was trying to be a bit more explicit about the types, to still make it clearer that Filter has child FilterGroup etc.

I could have done something like this

`SequenceEntry<Guid, SequenceEntry<Guid, Guid>>` for a filter id with child filter group id's that in turn have child filter item id's, and

`SequenceEntry<Guid, Guid>` for an indicator group id that has child indicator id's.

What I've ended up going with after changing to record types is:

```
    public abstract record SequenceEntry<TEntry, TChild>(TEntry Id, List<TChild> ChildSequence);

    public record FilterSequenceEntry(Guid Id, List<FilterGroupSequenceEntry> ChildSequence) :
        SequenceEntry<Guid, FilterGroupSequenceEntry>(Id, ChildSequence);

    public record FilterGroupSequenceEntry(Guid Id, List<Guid> ChildSequence) :
        SequenceEntry<Guid, Guid>(Id, ChildSequence);

    public record IndicatorGroupSequenceEntry(Guid Id, List<Guid> ChildSequence) :
        SequenceEntry<Guid, Guid>(Id, ChildSequence);
```